receiver/awscontainerinsightreceiver/internal/ecsInfo/ecs_instance_info.go

```diff
@@ -0,0 +1,128 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ecsinfo
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/awscontainerinsightreceiver/internal/host"
+)
+
+type containerInstanceInfoProvider interface {
+	GetClusterName() string
+	GetContainerInstanceID() string
+}
+
+type Requester interface {
+	Request(ctx context.Context, path string, logger *zap.Logger) ([]byte, error)
+}
+
+type hostIPProvider interface {
+	GetInstanceIP() string
+}
+
+type containerInstanceInfo struct {
+	logger                   *zap.Logger
+	httpClient               Requester
+	refreshInterval          time.Duration
+	ecsAgentEndpointProvider hostIPProvider
+	clusterName              string
+	containerInstanceID      string
+	readyC                   chan bool
+	sync.RWMutex
+}
+
+type ContainerInstance struct {
+	Cluster              string
+	ContainerInstanceArn string
+}
+
+type ecsInstanceInfoOption func(*containerInstanceInfo)
+
+func newECSInstanceInfo(ctx context.Context, ecsAgentEndpointProvider hostIPProvider,
+	refreshInterval time.Duration, logger *zap.Logger, httpClient Requester, readyC chan bool, options ...ecsInstanceInfoOption) containerInstanceInfoProvider {
+	cii := &containerInstanceInfo{
+		logger:                   logger,
+		httpClient:               httpClient,
+		refreshInterval:          refreshInterval,
+		ecsAgentEndpointProvider: ecsAgentEndpointProvider,
+		readyC:                   readyC,
+	}
+
+	for _, opt := range options {
+		opt(cii)
+	}
+
+	shouldRefresh := func() bool {
+		//stop the refresh once we get instance ID and type successfully
+		return cii.clusterName == "" || cii.containerInstanceID == ""
+	}
+	go host.RefreshUntil(ctx, cii.refresh, cii.refreshInterval, shouldRefresh, 0)
+
+	return cii
+}
+
+func (cii *containerInstanceInfo) refresh(ctx context.Context) {
+	containerInstance := &ContainerInstance{}
+	cii.logger.Info("Fetch instance id and type from ec2 metadata")
+	resp, err := cii.httpClient.Request(ctx, cii.getECSAgentEndpoint(), cii.logger)
+	if err != nil {
+		cii.logger.Warn("Failed to call ecsagent endpoint, error: ", zap.Error(err))
+	}
+
+	err = json.Unmarshal(resp, containerInstance)
+	if err != nil {
+		cii.logger.Warn("Failed to unmarshal: ", zap.Error(err))
+		cii.logger.Warn("Resp content is " + string(resp))
+	}
+
+	cluster := containerInstance.Cluster
+	instanceID, err := GetContainerInstanceIDFromArn(containerInstance.ContainerInstanceArn)
+
+	if err != nil {
+		cii.logger.Warn("Failed to get instance id from arn, error: ", zap.Error(err))
+	}
+
+	cii.Lock()
+	cii.clusterName = cluster
+	cii.containerInstanceID = instanceID
+	cii.Unlock()
+
+	if cii.clusterName != "" && cii.containerInstanceID != "" {
+		close(cii.readyC)
```

Comment:
What ensures this isn't closed twice?  Is there any way to eliminate this channel?  It seems to only be used for tests, which is a nasty code smell.

Reply:
Will add logic to check if the channel is closed or not